clippy_lints/src/empty_with_brackets.rs

```diff
@@ -97,22 +113,63 @@ impl EarlyLintPass for EmptyWithBrackets {
         }
     }
 
-    fn check_variant(&mut self, cx: &EarlyContext<'_>, variant: &Variant) {
+    fn check_variant(&mut self, cx: &LateContext<'_>, variant: &Variant<'_>) {
+        // Don't lint pub enums
+        if cx.effective_visibilities.is_reachable(variant.def_id) {
+            return;
+        }
+
         let span_after_ident = variant.span.with_lo(variant.ident.span.hi());
 
-        if has_brackets(&variant.data) && has_no_fields(cx, &variant.data, span_after_ident) {
+        if has_no_fields(cx, &variant.data, span_after_ident) {
+            match variant.data {
+                VariantData::Struct { .. } => {
+                    span_lint_and_then(
+                        cx,
+                        EMPTY_ENUM_VARIANTS_WITH_BRACKETS,
+                        span_after_ident,
+                        "enum variant has empty brackets",
+                        |diagnostic| {
+                            diagnostic.span_suggestion_hidden(
+                                span_after_ident,
+                                "remove the brackets",
+                                "",
+                                Applicability::MaybeIncorrect,
+                            );
+                        },
+                    );
+                },
+                VariantData::Tuple(..) => {
+                    if let Some(local_def_id) = variant.data.ctor_def_id()
+                        && !self.empty_tuple_enum_variants.contains_key(&local_def_id.to_def_id())
+                    {
+                        self.empty_tuple_enum_variants
+                            .insert(local_def_id.to_def_id(), Usage::Unused(span_after_ident));
+                    }
+                },
+                VariantData::Unit(..) => {},
+            }
+        }
+    }
+
+    fn check_expr(&mut self, _cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(def_id) = check_expr_for_enum_as_function(expr) {
+            self.empty_tuple_enum_variants.insert(def_id, Usage::Used);
+        }
+    }
+
+    fn check_crate_post(&mut self, cx: &LateContext<'_>) {
+        for &span in self.empty_tuple_enum_variants.values().filter_map(|v| match v {
+            Usage::Unused(span) => Some(span),
+            Usage::Used => None,
+        }) {
             span_lint_and_then(
```

Comment:
This needs to use `span_lint_hir_and_then` with the `HirId` of the variant now that we're delaying it to post-crate. Otherwise `#[allow]` attributes on the variant don't work. (Would also be a good test)